From efadf481eb0eac66a03988b6885e26df8166a0f6 Mon Sep 17 00:00:00 2001
From: Marco Martin <notmart@gmail.com>
Date: Fri, 26 May 2017 16:50:10 +0200
Subject: [PATCH 1/1] Set the QMenu's window transient parent to the
 causedPopup's window

Set the causing widget as the transient parent of the QMenu,
as on Wayland popup surfaces always need a transient parent set
to another surface always currently visible (being their menubar,
parent popup etc

QTBUG-60932

Change-Id: I503ed9d28d9a9b4bd70fc3221c0a92cdc923fdc2
---
 src/widgets/widgets/qmenu.cpp | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/widgets/widgets/qmenu.cpp b/src/widgets/widgets/qmenu.cpp
index 75704f7..6c89e77 100644
--- a/src/widgets/widgets/qmenu.cpp
+++ b/src/widgets/widgets/qmenu.cpp
@@ -2303,6 +2303,14 @@ QSize QMenu::sizeHint() const
 void QMenu::popup(const QPoint &p, QAction *atAction)
 {
     Q_D(QMenu);
+    /*Set the causing widget as the transient parent of the QMenu,
+     as on Wayland popup surfaces always need a transient parent set
+     to another surface always currently visible (being their menubar,
+     parent popup etc*/
+    if (winId() && d->causedPopup.widget && d->causedPopup.widget->isVisible() && d->causedPopup.widget->winId()) {
+        windowHandle()->setTransientParent(d->causedPopup.widget->windowHandle());
+    }
+
     if (d->scroll) { // reset scroll state from last popup
         if (d->scroll->scrollOffset)
             d->itemsDirty = 1; // sizeHint will be incorrect if there is previous scroll
-- 
2.7.4
